src/Features/Core/Portable/ConvertLinq/AbstractConvertLinqProvider.cs

```diff
@@ -0,0 +1,126 @@
+            protected readonly CancellationToken _cancellationToken;
+
+            public AnalyzerBase(SemanticModel semanticModel, CancellationToken cancellationToken)
+            {
+                _semanticModel = semanticModel;
+                _cancellationToken = cancellationToken;
+            }
+
+            public async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+            {
+                var document = context.Document;
+                var cancellationToken = context.CancellationToken;
+                var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                var nodeToRefactor = FindNodeToRefactor(root, context);
+                if (nodeToRefactor == null)
+                {
+                    return;
+                }
+
+                if (nodeToRefactor.ContainsDiagnostics)
+                {
+                    return;
+                }
+
+                if (TryConvert(nodeToRefactor, out var result))
+                {
+                    context.RegisterRefactoring(new MyCodeAction(Title, c => UpdateDocumentAsync(root, document, result)));
+                }
+            }
+
+            protected abstract bool TryConvert(TRefactor refactor, out DocumentUpdate documentUpdate);
+
+            protected virtual TRefactor FindNodeToRefactor(SyntaxNode root, CodeRefactoringContext context) =>
+                root.FindNode(context.Span) as TRefactor ?? default;
+
+            protected abstract string Title { get; }
+
+            private Task<Document> UpdateDocumentAsync(
+                SyntaxNode root,
+                Document document,
+                DocumentUpdate documentUpdate)
+            {
+                root = documentUpdate.UpdateRoot(root);
+                return Task.FromResult(document.WithSyntaxRoot(root));
+            }
+        }
+
+        protected sealed class DocumentUpdate
+        {
+            private ImmutableArray<(SyntaxNode source, ImmutableArray<SyntaxNode> destinations)> _updates;
+
+            public DocumentUpdate(SyntaxNode source, SyntaxNode destination)
+                => _updates = ImmutableArray.Create((source, ImmutableArray.Create(destination)));
+
+            public DocumentUpdate(SyntaxNode source, IEnumerable<SyntaxNode> destination)
+                => _updates = ImmutableArray.Create((source, ImmutableArray.CreateRange(destination)));
+
+            public SyntaxNode UpdateRoot(SyntaxNode root)
+            {
+                foreach (var updateItem in _updates)
+                {
+                    if (updateItem.destinations.Any())
+                    {
+                        root = root.ReplaceNode(updateItem.source, updateItem.destinations.Select(node => node.WithAdditionalAnnotations(Simplifier.Annotation)));
+                    }
+                    else
+                    {
+                        root = root.RemoveNode(updateItem.source, SyntaxRemoveOptions.AddElasticMarker);
```

Comment:
i don't see how this coudl work.  after you replace a node and get a new root, there's no way you'd be able to call .ReplaceNode again with a node from the original tree.  The original node won't be found in the new root.

Note: this is why SyntaxEditor exists.  So you can make multiple edits without having to do the work to track these sorts of things.

Reply:
+1 for using SyntaxEditor here.

---
In reply to: [176296750](https://github.com/dotnet/roslyn/pull/25362#discussion_r176296750) [](ancestors = 176296750)

Reply:
Thank you! I do not need remove for now, and likely in other linq to * refactorings. Removing this and simplifying DocumentUpdate.

Consider the replacement with SyntaxEditor in a separate PR. Handled in [https://github.com/dotnet/roslyn/issues/25639](https://github.com/dotnet/roslyn/issues/25639)

---
In reply to: [176534999](https://github.com/dotnet/roslyn/pull/25362#discussion_r176534999) [](ancestors = 176534999,176296750)

Reply:
Note: my comment is about *both* the remove and the replace.  The way you've structured your code, it will not work if you need to make multiple changes.  i.e. if you expect the above code to properly make two replaces, it will not.  Either you need to make a single update, in which case the code shoudl reflect it.  Or, you need to make multiple updates, in which case you either need to do a bottom up tree walk (not recommended), or use SyntaxEditor.

Reply:
for now, there is a need for a single update. I will replace it with syntax editor in the next PR.

---
In reply to: [176577975](https://github.com/dotnet/roslyn/pull/25362#discussion_r176577975) [](ancestors = 176577975)